source/stun_serializer.c

```diff
@@ -0,0 +1,860 @@
+                                               STUN_ATTRIBUTE_TYPE_XOR_RELAYED_ADDRESS );
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_GetIntegrityBuffer( StunContext_t * pCtx,
+                                                uint8_t ** ppStunMessage,
+                                                uint16_t * pStunMessageLength )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pCtx == NULL ) ||
+        ( pStunMessageLength == NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if ( pCtx->pStart != NULL )
+        {
+            // Fix-up the packet length with message integrity and without the STUN header
+            STUN_WRITE_UINT16( ( uint8_t * ) &( pCtx->pStart[ STUN_HEADER_MESSAGE_LENGTH_OFFSET ] ),
+                                pCtx->currentIndex - STUN_HEADER_LENGTH + STUN_ATTRIBUTE_TOTAL_LENGTH( STUN_HMAC_VALUE_LENGTH ));
+
+            *ppStunMessage =  ( uint8_t * ) (pCtx->pStart);
+        }
+
+        *pStunMessageLength = pCtx->currentIndex;
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_GetFingerprintBuffer( StunContext_t * pCtx,
+                                                  uint8_t ** ppStunMessage,
+                                                  uint16_t * pStunMessageLength )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pCtx == NULL ) ||
+        ( pStunMessageLength == NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if ( pCtx->pStart != NULL )
+        {
+            // Fix-up the packet length with fingerprint CRC and without the STUN header
+            STUN_WRITE_UINT16( ( uint8_t * ) &( pCtx->pStart[ STUN_HEADER_MESSAGE_LENGTH_OFFSET ] ),
+                                pCtx->currentIndex - STUN_HEADER_LENGTH + STUN_ATTRIBUTE_TOTAL_LENGTH( STUN_ATTRIBUTE_FINGERPRINT_LENGTH ));
+
+            *ppStunMessage =  ( uint8_t * ) (pCtx->pStart);
+        }
+
+        *pStunMessageLength = pCtx->currentIndex;
+    }
+
+    return result;
+}
+/*-----------------------------------------------------------*/
+
+StunResult_t StunSerializer_Finalize( StunContext_t * pCtx,
+                                      uint32_t * pStunMessageLength )
+{
+    StunResult_t result = STUN_RESULT_OK;
+
+    if( ( pCtx == NULL ) ||
+        ( pStunMessageLength == NULL ) )
+    {
+        result = STUN_RESULT_BAD_PARAM;
+    }
+
+    if( result == STUN_RESULT_OK )
+    {
+        if ( pCtx->pStart != NULL )
+        {
+            /* Update the message length field in the header. */
+            STUN_WRITE_UINT16( ( uint8_t * ) &( pCtx->pStart[ STUN_HEADER_MESSAGE_LENGTH_OFFSET ] ),
+                                pCtx->currentIndex - STUN_HEADER_LENGTH );
+        }
+        *pStunMessageLength = pCtx->currentIndex;
+
```

Comment:
In APIs like `StunSerializer_GetFingerprintBuffer`, pCtx->currentIndex is put in a uint16_t variable, whereas here it is put to a uint32_t variable. What is the reason behind this?

Reply:
Changed it uniformly to size_t